poethepoet/task/shell.py

```diff
@@ -31,42 +42,120 @@ def _handle_run(
-            "Couldn't locate bash executable to run shell task. Installing WSL should "
-            "fix this."
+    def _get_interpreter_config(self) -> Tuple[str, ...]:
+        result: Union[str, Tuple[str, ...]] = self.options.get(
+            "interpreter", self._config.shell_interpreter
         )
+        if isinstance(result, str):
+            return (result,)
+        return tuple(result)
+
+    def resolve_interpreter_cmd(self) -> Optional[List[str]]:
+        """
+        Return a formatted command for the first specified interpreter that can be
+        located.
+        """
+        for item in self._get_interpreter_config():
+            executable = self._locate_interpreter(item)
+            if executable is None:
+                return None
+            if item in ("pwsh", "powershell"):
+                return [executable, "-NoLogo", "-Command", "-"]
+            return [executable]
+        return None
+
+    def _locate_interpreter(self, interpreter: str) -> Optional[str]:
+        result = None
+        prog_files = environ.get("PROGRAMFILES", "C:\\Program Files")
+
+        # Try use $SHELL from the environment as a hint
+        shell_var = environ.get("SHELL", "")
+        if shell_var.endswith(f"/{interpreter}") and which(shell_var) == shell_var:
+            result = shell_var
+
+        elif interpreter == "posix":
+            # look for any known posix shell
+            result = (
+                self._locate_interpreter("sh")
+                or self._locate_interpreter("bash")
+                or self._locate_interpreter("zsh")
+            )
+
+        elif interpreter == "sh":
+            result = which("sh") or which("/bin/sh")
+
+        elif interpreter == "bash":
+            result = which("bash") or which("/bin/bash")
+
+            # Specifically look for git bash on windows
+            if result is None and self._is_windows:
+                result = which(f"{prog_files}\\Git\\bin\\bash.exe")
+
+        elif interpreter == "zsh":
+            result = which("zsh") or which("/bin/zsh")
+
+        elif interpreter == "fish":
+            result = which("fish") or which("/bin/fish")
+
+        elif interpreter in ("pwsh", "powershell"):
+            # Look for the pwsh executable and verify the version matches
+            result = (
+                which("pwsh")
+                or which(f"{prog_files}\\PowerShell\\7\\pwsh.exe")
+                or which(f"{prog_files}\\PowerShell\\6\\pwsh.exe")
+            )
+
+            if result is None and interpreter == "powershell" and self._is_windows:
+                # Look for older versions of powershell
+                result = which("powershell") or which(
+                    environ.get("WINDIR", "C:\\Windows")
+                    + "\\System32\\WindowsPowerShell\\v1.0\\powershell.EXE"
+                )
+
+        elif interpreter == "python":
+            result = which("python") or which("python3") or sys.executable
```

Comment:
I'm curious under what circumstances one wouldn't use `sys.executable` when running a python script?

Reply:
I hope it's never needed, in that python or python3 should be on the path in any sane environment, but if somehow they're not then `sys.executable` gives us the one thing we know about for sure which is the python interpreter of the current process.

Reply:
Forgive me if I am wrong, but I think @TBBle may have meant that using `sys.executable` is the most desirable option, rather than the last thing in the `or` chain.

Reply:
I see, maybe I misread. That is actually what I thought originally when implementing the script task type, but then [I changed my mind about it](https://github.com/nat-n/poethepoet/commit/fd85005c2062eac7863d00161aad408966fb8a0f#diff-6530823c1ef123e90f1bc9301d47e75c652a895a72e64f5f4e14f891e25c3f3aR26) for reasons that I can't recall with confidence.

However my understanding at this point is that given the many ways that python versions can be installed and managed on a system, it's a safer bet to use whatever is configured in the current environment rather than whatever the poe executable is using. 

For example poe might have been installed with the system python, but the task is being executed within a virtual env (perhaps activated by the PoetryExecutor or VirtualenvExecutor inside poe) which is probably more relevant and may reference a different python version (and site_packages) than the poe CLI is using.

Reply:
Ah, that's a good point. I've only used Poe installed via Poetry as a task runner, and did not consider a system-installed Poe.